File: python/241.different-ways-to-add-parentheses.py


class Solution:
    def diffWaysToCompute(self, expression: str) -> list[int]:

        n = len(expression)
        dp: list[list[list[int]]] = [[[] for _ in range(n)] for _ in range(n)]
        self._init_base_cases(expression, dp)

        for length in range(3, n + 1):
            for start in range(n - length + 1):
                self._process_expression(expression, start, start + length - 1, dp)

        return dp[0][n - 1]

    def _init_base_cases(self, expression: str, dp: list[list[list[int]]]) -> None:
        for i in range(len(expression)):
            if expression[i].isdigit():
                dp[i][i].append(int(expression[i]))

                if i + 1 < len(expression) and expression[i + 1].isdigit():
                    dp[i][i + 1].append(int(expression[i: i + 2]))

    def _process_expression(
        self, expression: str, start: int, end: int, dp: list[list[list[int]]]
    ) -> None:
        for split in range(start, end + 1):
            if expression[split].isdigit():
                continue

            leftResult = dp[start][split - 1]
            rightResult = dp[split + 1][end]

            dp[start][end].extend(
                self._compute(expression[split], leftResult, rightResult)
            )

    def _compute(self, op: str, left: list[int], right: list[int]) -> list[int]:

        results: list[int] = []
        for leftNum in left:
            for rightNum in right:
                if op == "+":
                    results.append(leftNum + rightNum)
                elif op == "-":
                    results.append(leftNum - rightNum)
                elif op == "*":
                    results.append(leftNum * rightNum)

        return results

File: golang/241.different-ways-to-add-parentheses.go
func diffWaysToCompute(expression string) []int {
	if len(expression) == 0 {
		return []int{}
	}

	if len(expression) == 1 && '0' <= expression[0] && expression[0] <= '9' {
		return []int{int(expression[0] - '0')}
	}

	if len(expression) == 2 && '0' <= expression[0] && expression[0] <= '9' && '0' <= expression[1] && expression[1] <= '9' {
		return []int{int(expression[0]-'0')*10 + int(expression[1]-'0')}
	}

	result := make([]int, 0)
	for i := 0; i < len(expression); i++ {
		if '0' <= expression[i] && expression[i] <= '9' {
			continue
		}

		leftResult := diffWaysToCompute(expression[:i])
		rightResult := diffWaysToCompute(expression[i+1:])

		for _, left := range leftResult {
			for _, right := range rightResult {
				if expression[i] == '+' {
					result = append(result, left+right)
				} else if expression[i] == '-' {
					result = append(result, left-right)
				} else if expression[i] == '*' {
					result = append(result, left*right)
				}
			}
		}

	}

	return result
}
